File: demo-task-executor-java/src/test/java/demo/executor/TaskTest.java
package demo.executor;

import org.testng.annotations.Test;

import java.util.HashSet;
import java.util.Set;

import static org.testng.Assert.assertEquals;
import static org.testng.Assert.assertTrue;
import static org.testng.Assert.fail;

/**
 * Unit tests for {@link demo.executor.Task}.
 */
@Test
public class TaskTest extends AbstractConcurrencyTest {

    private static final String RESULT = "result";
    private static final RuntimeException EXCEPTION = new RuntimeException();

    @Test
    public void testSetResult() {
        Set<String> consumer = new HashSet<>();
        Task<String> task = getTestTaskWithResult(consumer);
        assertEquals(task.isFinished(), false);

        task.setResult(RESULT);

        assertEquals(task.isFinished(), true);
        assertEquals(consumer.size(), 1);
        assertTrue(consumer.contains(RESULT));
    }

    @Test
    public void testSetException() {
        Set<Exception> consumer = new HashSet<>();
        Task<String> task = getTaskWithException(consumer);
        assertEquals(task.isFinished(), false);

        task.setException(EXCEPTION);

        assertEquals(task.isFinished(), true);
        assertEquals(consumer.size(), 1);
        assertTrue(consumer.contains(EXCEPTION));
    }

    @Test
    public void testGetCallableWithResult() throws Exception {
        Task<String> task = getTestTaskWithResult(new HashSet<>());
        assertEquals(task.getCallable().call(), RESULT);
    }

    @Test(expectedExceptions = RuntimeException.class)
    public void testGetCallableWithException() throws Exception {
        Task<String> task = getTaskWithException(new HashSet<>());
        task.getCallable().call();
    }

    @Test(expectedExceptions = IllegalStateException.class)
    public void testMarkFinishedException() {
        Task<String> task = getTestTaskWithResult(new HashSet<>());
        task.setResult(RESULT);
        task.markFinished();
    }

    private Task<String> getTestTaskWithResult(Set<String> consumer) {
        return new Task<>("task with result", String.class,
                () -> {
                    return RESULT;
                },
                (result, exception) -> {
                    if (exception.isPresent()) {
                        fail();
                    } else {
                        consumer.add(result.get());
                    }
                }
        );
    }

    private Task<String> getTaskWithException(Set<Exception> consumer) {
        return new Task<>("task with exception", String.class,
                () -> {
                    throw EXCEPTION;
                },
                (result, exception) -> {
                    if (exception.isPresent()) {
                        consumer.add(exception.get());
                    } else {
                        fail();
                    }
                }
        );
    }
}

File: demo-task-executor-kotlin/src/test/kotlin/demo/executor/TaskTest.kt
package demo.executor

import org.testng.Assert.assertEquals
import org.testng.Assert.assertTrue
import org.testng.Assert.fail
import org.testng.annotations.Test
import java.util.HashSet
import java.util.concurrent.Callable

/**
 * Unit tests for [demo.executor.Task].
 */
@Test
class TaskTest : AbstractConcurrencyTest() {

    @Test
    fun testSetResult() {
        val consumer = HashSet<String>()
        val task = getTestTaskWithResult(consumer)
        assertEquals(task.isFinished(), false)

        task.setResult(RESULT)

        assertEquals(task.isFinished(), true)
        assertEquals(consumer.size, 1)
        assertTrue(consumer.contains(RESULT))
    }

    @Test
    fun testSetException() {
        val consumer = HashSet<Exception>()
        val task = getTaskWithException(consumer)
        assertEquals(task.isFinished(), false)

        task.setException(EXCEPTION)

        assertEquals(task.isFinished(), true)
        assertEquals(consumer.size, 1)
        assertTrue(consumer.contains(EXCEPTION))
    }

    @Test
    fun testGetCallableWithResult() {
        val task = getTestTaskWithResult(HashSet<String>())
        assertEquals(task.getCallable().call(), RESULT)
    }

    @Test(expectedExceptions = arrayOf(RuntimeException::class))
    fun testGetCallableWithException() {
        val task = getTaskWithException(HashSet<Exception>())
        task.getCallable().call()
    }

    @Test(expectedExceptions = arrayOf(IllegalStateException::class))
    fun testMarkFinishedException() {
        val task = getTestTaskWithResult(HashSet<String>())
        task.setResult(RESULT)
        task.markFinished()
    }

    private fun getTestTaskWithResult(consumer: MutableSet<String>): Task<String> {
        return Task("task with result", String::class,
                Callable { RESULT }
        ) { result, exception ->
            if (exception.isPresent) {
                fail()
            } else {
                consumer.add(result.get())
            }
        }
    }

    private fun getTaskWithException(consumer: MutableSet<Exception>): Task<String> {
        return Task("task with exception", String::class,
                Callable<String> { throw EXCEPTION }
        ) { result, exception ->
            if (exception.isPresent) {
                consumer.add(exception.get())
            } else {
                fail()
            }
        }
    }

    companion object {

        private const val RESULT = "result"
        private val EXCEPTION = RuntimeException()
    }
}
